runtime/compiler/control/CompilationThread.cpp

```diff
@@ -5213,169 +5207,209 @@ TR::CompilationInfo::getNextMethodToBeCompiled(TR::CompilationInfoPerThread *com
+   // split into two parts. In the event we are a diagnostic thread, the JitDump process will have ensured the method
+   // compilation queue has been purged and all further compilations have been suspended until the entire JitDump
+   // process in complete. This also means if we are a diagnostic thread, then the only entries in the queue should
+   // be JitDump compile requests. We will ensure this is the case via a fatal assert.
+   //
+   // In addition there can be scenarios in which a non-diagnostic compilation thread is still attempting to process
+   // entries while the JitDump process (in a crashed thread) is happening. This is because one compilation thread
+   // must always be active, and there is a timing hole between purging of the queue in the JitDump process, and when
+   // the diagnostic thread is resumed. This means a non-diagnostic thread could attempt to pick up a JitDump
+   // compilation request (see eclipse/openj9#11772 for details). We must ensure that this does not happen and simply
+   // skip the request if we are a non-diagnostic thread attempting to process a JitDump compilation.
+   if (compInfoPT->isDiagnosticThread())
       {
-      // If the request is sync or AOT load or InstantReplay, take it now
-      if (compInfoPT->isDiagnosticThread() // InstantReplay compilations must be processed immediately
-          || _methodQueue->_priority >= CP_SYNC_MIN // sync comp
-          || _methodQueue->_methodIsInSharedCache == TR_yes // very cheap relocation
-#if defined(J9VM_OPT_JITSERVER)
-          || getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER // compile right away in server mode
-#endif
-         )
-         {
-         nextMethodToBeCompiled = _methodQueue;
-         _methodQueue = _methodQueue->_next;
-         }
-      // Check if we need to throttle
-      else if (exceedsCompCpuEntitlement() == TR_yes &&
-              !compThreadCameOutOfSleep && // Don't throttle a comp thread that has just slept its share of time
-              (TR::Options::_compThreadCPUEntitlement < 100 || getNumCompThreadsActive() * 100 > (TR::Options::_compThreadCPUEntitlement + 50)))
-         {
-         // If at all possible suspend the compilation thread
-         // Otherwise, perform a timed wait
-         if (getNumCompThreadsActive() > 1)
-            *compThreadAction = SUSPEND_COMP_THREAD_EXCEED_CPU_ENTITLEMENT;
-         else
-            *compThreadAction = THROTTLE_COMP_THREAD_EXCEED_CPU_ENTITLEMENT;
-         }
-      // Avoid two concurrent hot compilations
-      else if (getNumCompThreadsCompilingHotterMethods() <= 0 || // no hot compilation in progress
-               _methodQueue->_weight < TR::Options::_expensiveCompWeight) // This is a cheaper comp
+      // We never want to self-suspend the diagnostic thread. The JitDump process will do it after it is complete.
+      *compThreadAction = GO_TO_SLEEP_EMPTY_QUEUE;
```

Comment:
I don't see any code in JitDump.cpp that suspends the diagnostic thread; I guess we leave it for `JitShutdown` to do so?

Reply:
See above explanation for what happens after the JitDump process completes. The JVM will just terminate via `exit(-1)` so this thread does not need to be shut down. It is only ever resurrected during a crash operation which is _not_ recoverable.